Sources/SwiftCoAP/SCMessage.swift

```diff
@@ -32,87 +34,171 @@ protocol SCCoAPTransportLayerDelegate: class {
-        } catch {
-            return false
+
+    private func mustGetConnection(forHost host: String, port: UInt16) -> NWConnection {
+        os_log("Currently %d NWConnection object(s) are alive", connections.count)
+        os_log("Getting connection object for HOST %@, PORT %d", log: .default, type: .info, host, port)
+        let connectionKey = HostPortKey(host: host, port: port)
+        if let connection = connections[connectionKey], connection.state != .cancelled {
+            os_log("Reusing existing NWConnection for HOST %@, PORT %d", log: .default, type: .info, host, port)
+            return connection
+        }
+        let connection = NWConnection(host: NWEndpoint.Host(host), port: NWEndpoint.Port(rawValue: port)!, using: .udp)
+        connections[connectionKey] = connection
+        os_log("New NWConnection created for HOST %@, PORT %d", log: .default, type: .info, host, port)
+        return connection
+    }
+
+    private func startReads(from connection: NWConnection, withHostPort hostPort: HostPortKey) {
+        var waitingForNextDatagram = true
+        while connection.state == .ready {
+            if waitingForNextDatagram {
+                waitingForNextDatagram = false
+                connection.receiveMessage { [weak self] data, context, complete, error in
+                    waitingForNextDatagram = true
+                    guard let self = self else { return }
+                    if error != nil {
+                        self.transportLayerDelegate.transportLayerObject(self, didFailWithError: error! as NSError)
+                        return
+                    }
+                    if let data = data {
+                        self.transportLayerDelegate.transportLayerObject(self, didReceiveData: data, fromHost: hostPort.host, port: hostPort.port)
+                    }
+                }
+            }
         }
-        return true
     }
 }
 
 extension SCCoAPUDPTransportLayer: SCCoAPTransportLayerProtocol {
-    func sendCoAPData(_ data: Data, toHost host: String, port: UInt16) throws {
-        try startListening()
-        udpSocket.send(data, toHost: host, port: port, withTimeout: 0, tag: udpSocketTag)
-        udpSocketTag = (udpSocketTag % Int.max) + 1
-    }
-    
-    func closeTransmission() {
-        udpSocket.close()
-        udpSocket = nil
+    public func sendCoAPData(_ data: Data, toHost host: String, port: UInt16) throws {
+        var connection = mustGetConnection(forHost: host, port: port)
+        let hostPort = HostPortKey(host: host, port: port)
+        if connection.stateUpdateHandler == nil {
+            connection = setupStateUpdateHandler(for: connection, withHostPort: hostPort)
+        }
+        if connection.state == .setup {
+            connection.start(queue: DispatchQueue.global(qos: .utility))
+        }
+        connection.send(content: data, completion: .contentProcessed{ [weak self] error in
+            guard let self = self else { return }
```

Comment:
If we make the delegate weak we don't really need to do this dance, can just do `self?.transportLayerDelegate?.transportLayerObject(…)`

Reply:
The delegate is weak already. This guard is needed to pass the `self` (which is the implementation of `SCCoAPTransportLayerProtocol` itself) to delegate's `transportLayerObject` method. Why so, only original author knows, or, maybe we'll get it reimplementing/overriding `SCClient` or `SCServer` methods. Leave it unchanged for now.

Reply:
Oh, right, that makes sense.